### PR TITLE
Returning at most countIdsPerSoql records from startFetch / continueFetch

### PR DESCRIPTION
**Pull request by wmathurin**
To avoid too much memory usage.

If you have
```
typejA -> idA1 ... idAn
typeB -> idB1 ... idBn
etc
```
We are building a list that looks like `[idA1 ... idAn idB1 ... idBn ... ]`
And chunking that list, so if there are a lot of A's, we might do only A's in the first round, but if there are less than 500 A's we might do all the A's and some B's etc - but always no more than 500 records.

The new class `TypedIds` is a helper class that has methods for "chunking" and also for going back to the map of object type to list of ids.

### CODE REVIEW

File: libs/MobileSync/src/com/salesforce/androidsdk/mobilesync/target/BriefcaseSyncDownTarget.java [hunk 1]
```diff
@@ -174,6 +180,16 @@ public int cleanGhosts(SyncManager syncManager, String soupName, long syncId) th
         return countGhosts;
     }
 
+    @Override
```
Comment (wmathurin) [inline libs/MobileSync/src/com/salesforce/androidsdk/mobilesync/target/BriefcaseSyncDownTarget.java:183]: This is unrelated.
Without overriding this method a sync down with LEAVE_IF_CHANGED will not work correctly if the briefcase targets multiple object types.
(Need to add a test for that also).

**Review by bhariharan — APPROVED** [review]
